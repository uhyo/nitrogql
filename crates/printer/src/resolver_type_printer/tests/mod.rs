#![cfg(test)]

use insta::assert_snapshot;
use nitrogql_ast::TypeSystemDocument;
use nitrogql_parser::parse_type_system_document;
use nitrogql_semantics::resolve_extensions;
use sourcemap_writer::JustWriter;

use crate::ResolverTypePrinterPlugin;

use super::{
    error::ResolverTypePrinterResult, options::ResolverTypePrinterOptions,
    printer::ResolverTypePrinter,
};

struct DummyPlugin;
impl ResolverTypePrinterPlugin for DummyPlugin {
    fn transform_resolver_output_types<'src>(
        &self,
        _document: &TypeSystemDocument<'src>,
        _options: &ResolverTypePrinterOptions,
        _base: std::collections::HashMap<&'src str, crate::ts_types::TSType>,
    ) -> std::collections::HashMap<&'src str, crate::ts_types::TSType> {
        unimplemented!()
    }
    fn transform_document_for_resolvers<'src>(
        &self,
        _document: &TypeSystemDocument<'src>,
    ) -> TypeSystemDocument<'src> {
        unimplemented!()
    }
}

static EMPTY_PLUGINS: &[DummyPlugin] = &[];

#[test]
fn resolver_printing() {
    let doc = parse_type_system_document(
        "
            type User implements HasID {
                id: ID!
                name: String!
                type: UserType!
                age: Int
                posts: [Post!]!
            }
            interface HasID {
                id: ID!
            }
            enum UserType {
                NormalUser
                PremiumUser
            }
            type Bot implements HasID {
                id: ID!
            }

            union UserOrBot = User | Bot

            type Post {
                id: ID!
                title: String!
                tags: [String!]
                body: String!
            }

            input UserSearchQuery {
                age: Int
                name: String
                keywords: [String!]
            }

            type Query {
                me: User!
            }
            ",
    )
    .unwrap();
    let doc = resolve_extensions(doc).unwrap();
    let printed = print_document(
        &doc,
        ResolverTypePrinterOptions {
            schema_source: "schema".into(),
            ..Default::default()
        },
        EMPTY_PLUGINS,
    )
    .unwrap();
    assert_snapshot!(printed);
}

fn print_document(
    document: &TypeSystemDocument,
    options: ResolverTypePrinterOptions,
    plugins: &[impl ResolverTypePrinterPlugin],
) -> ResolverTypePrinterResult<String> {
    let mut result = String::new();
    let mut writer = JustWriter::new(&mut result);
    let mut printer = ResolverTypePrinter::new(options, &mut writer);
    printer.print_document(document, plugins)?;
    Ok(result)
}
