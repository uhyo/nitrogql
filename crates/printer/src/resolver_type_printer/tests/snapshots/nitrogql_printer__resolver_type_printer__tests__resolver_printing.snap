---
source: crates/printer/src/resolver_type_printer/tests/mod.rs
expression: printed
---
import type { GraphQLResolveInfo } from "graphql";
import type * as Schema from "schema";
type __Resolver<Parent, Args, Context, Result> = (parent: Parent, args: Args, context: Context, info: GraphQLResolveInfo) => Result | Promise<Result>;
type __TypeResolver<Obj, Context, Result> = (object: Obj, context: Context, info: GraphQLResolveInfo) => Result | Promise<Result>;
type User = Omit<Schema.User, "__typename">;
type Bot = Omit<Schema.Bot, "__typename">;
type Post = Omit<Schema.Post, "__typename">;
type Query = Omit<Schema.Query, "__typename">;
type HasID = User | Bot;
type UserOrBot = User | Bot;
type UserType = Schema.UserType;
type UserSearchQuery = Schema.UserSearchQuery;
export type Resolvers<Context> = {
  User: {
    id: __Resolver<User, {}, Context, ID>;
    name: __Resolver<User, {}, Context, String>;
    type: __Resolver<User, {}, Context, UserType>;
    age: __Resolver<User, {}, Context, Int | null>;
    posts: __Resolver<User, {}, Context, (Post)[]>;
  };
  Bot: {
    id: __Resolver<Bot, {}, Context, ID>;
  };
  Post: {
    id: __Resolver<Post, {}, Context, ID>;
    title: __Resolver<Post, {}, Context, String>;
    tags: __Resolver<Post, {}, Context, (String)[] | null>;
    body: __Resolver<Post, {}, Context, String>;
  };
  Query: {
    me: __Resolver<Query, {}, Context, User>;
  };
  HasID: {
    __resolveType: __TypeResolver<User | Bot, Context, "User" | "Bot">;
  };
  UserOrBot: {
    __resolveType: __TypeResolver<User | Bot, Context, "User" | "Bot">;
  };
};
export type ResolverOutput<T extends "User" | "Bot" | "Post" | "Query" | "HasID" | "UserOrBot" | "UserType" | "UserSearchQuery"> = 
{
  User: User;
  Bot: Bot;
  Post: Post;
  Query: Query;
  HasID: HasID;
  UserOrBot: UserOrBot;
  UserType: UserType;
  UserSearchQuery: UserSearchQuery;
}[T];

